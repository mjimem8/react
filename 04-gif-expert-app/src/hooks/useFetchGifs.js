import { useEffect, useState } from "react"
import { getGifs } from "../helpers/getGifs";

export const useFetchGifs = (category) => {
    
    const [state, setState] = useState({
        data: [],
        loading: true
    });
    
    //useEffect es muy util porque cuando utilizamos setCount por ejemplo
    //react ejecuta de nuevo todo el codigo del componente por lo cual podemos 
    //machacar valores que no queremos
    useEffect( () => {
        getGifs(category)
            .then(imgs => setState({
                data: imgs,
                loading: false
            }));
        //en el array señalamos las dependecias/logica por la cual
        //debe de ejecutarse getGifs
        //es decir sin dependencias getGifs se ejecutará solo una vez
        //pero hemos indicado que se ejecute la primera vez y cuando cambie de valor category
    }, [ category ]);

    return state;
}