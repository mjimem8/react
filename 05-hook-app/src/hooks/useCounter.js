import { useState } from 'react'

export const useCounter = (initialState = 10) => {
    const [state, setstate] = useState(initialState);

    const reset = () => {
        setstate(initialState);
    }

    const increment = () => {
        setstate(state + 1);
    }

    const decrement = () => {
        setstate(state - 1);
    }

    return { 
        counter: state,
        increment,
        decrement,
        reset
     };
}
