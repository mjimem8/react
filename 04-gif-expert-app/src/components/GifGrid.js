import React from 'react';
import PropTypes from 'prop-types';
import { useFetchGifs } from '../hooks/useFetchGifs';
import { GifGridItem } from './GifGridItem';

const GifGrid = ({category}) => {
    
    const { data: images, loading } = useFetchGifs(category);
    // const [count, setCount] = useState(0);

    return (
        <>
            <h3>{ category }</h3>

            {/* forma de mostrar loading con un if sin ser ternario */}
            { loading && <p>Loading</p> }

            <div className="card-grid">
                {/* <h3>{ count }</h3>
                <button onClick={ () => setCount(count + 1)}></button> */}
                    { 
                        images.map( img => (
                            <GifGridItem key={ img.id } {...img}/>
                        )) 
                    }
            </div>
        </>
    )
}

export default GifGrid

GifGrid.propTypes = {
    category: PropTypes.string.isRequired
}
