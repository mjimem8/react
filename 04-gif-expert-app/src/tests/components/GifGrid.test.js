import React from 'react';
import '@testing-library/jest-dom';
import { shallow } from "enzyme";
import GifGrid from "../../components/GifGrid";
import { useFetchGifs } from '../../hooks/useFetchGifs';
jest.mock('../../hooks/useFetchGifs');

describe('Pruebas en el <GifGrid />', () => {
    const category = 'One Punch';

    test('debe de mostrarse correctamente', () => {

        //moquea los datos que devuelve dicha funci√≥n
        useFetchGifs.mockReturnValue({
            data: [],
            loading: true
        });

        const wrapper = shallow(<GifGrid  category={category} />);

        expect(wrapper).toMatchSnapshot();
    });

    test('debe de mostrar items cuando se cargan imagenes useFetchGifs', () => {
        const imgs = [{
            id: 'ABC',
            url: 'https://localhost/cualquier/cosa.jpg',
            title: 'Cualquier cosa'
        }];
        
        useFetchGifs.mockReturnValue({
            data: imgs,
            loading: false
        });

        const wrapper = shallow(<GifGrid  category={category} />);

        expect(wrapper.find('p').exists()).toBe(false);
        expect(wrapper.find('GifGridItem').length).toBe(imgs.length);
    });


})