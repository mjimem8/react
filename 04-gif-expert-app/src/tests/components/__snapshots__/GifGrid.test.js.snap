// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en el <GifGrid /> debe de mostrarse correctamente 1`] = `
<Fragment>
  <h3>
    One Punch
  </h3>
  <p>
    Loading
  </p>
  <div
    className="card-grid"
  />
</Fragment>
`;
